x-pack/plugins/security_solution/common/endpoint/models/policy_config_helpers.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { PolicyConfig } from '../types';
+import { ProtectionModes } from '../types';
+
+/**
+ * Returns a copy of the passed `PolicyConfig` with all protections set to disabled.
+ *
+ * @param policy
+ * @returns
+ */
+export const disableProtections = (policy: PolicyConfig): PolicyConfig => {
```

Comment:
I was thinking about your comment on slack that `getCloudPolicyConfig` does not have all protections disabled, and I'm wondering how we can enforce that all protections will be disabled in case more protections are added in the future since protections are not included under a specific object.

So I think, instead of a `disabledProtections` function, we could create a `policyFactoryWithDisabledProtections`, then Typescript would ensure every new field added on PolicyConfig is explicitly set on the policy:

For example (in [policy_config.ts](https://github.com/elastic/kibana/blob/d2ea8bf58f68621147b0e7017b654674b9081f2d/x-pack/plugins/security_solution/common/endpoint/models/policy_config.ts)):

```js
/**
 * Return a policyFactory with all protections disabled`
 */
export const policyFactoryWithDisabledProtections = (): ReturnType<typeof policyFactory> => {
  const policy = policyFactory();
  return {
    windows: {
      events: {
        ...policy.windows.events
      },
      malware: {
        mode: ProtectionModes.off,
        blocklist: true,
      },
      ransomware: {
        mode: ProtectionModes.off,
        supported: false,
      },
      memory_protection: {
        mode: ProtectionModes.off,
        supported: false,
      },
      behavior_protection: {
        mode: ProtectionModes.off,
        supported: false,
      },
      popup: {
        malware: {
          message: '',
          enabled: false,
        },
        ransomware: {
          message: '',
          enabled: false,
        },
        memory_protection: {
          message: '',
          enabled: false,
        },
        behavior_protection: {
          message: '',
          enabled: false,
        },
      },
      logging: {
        ...policy.windows.logging
      },
      antivirus_registration: {
        enabled: false,
      },
      attack_surface_reduction: {
        credential_hardening: {
          enabled: false,
        },
      },
    },
    mac: {
      events: {
        ...policy.mac.events
      },
      malware: {
        mode: ProtectionModes.off,
        blocklist: true,
      },
      behavior_protection: {
        mode: ProtectionModes.off,
        supported: true,
      },
      memory_protection: {
        mode: ProtectionModes.off,
        supported: true,
      },
      popup: {
        malware: {
          message: '',
          enabled: false,
        },
        behavior_protection: {
          message: '',
          enabled: false,
        },
        memory_protection: {
          message: '',
          enabled: false,
        },
      },
      logging: {
        ...policy.mac.logging
      },
    },
    linux: {
      events: {
        ...policy.linux.events
      },
      malware: {
        mode: ProtectionModes.off,
        blocklist: false,
      },
      behavior_protection: {
        mode: ProtectionModes.off,
        supported: false,
      },
      memory_protection: {
        mode: ProtectionModes.off,
        supported: false,
      },
      popup: {
        malware: {
          message: '',
          enabled: false,
        },
        behavior_protection: {
          message: '',
          enabled: false,
        },
        memory_protection: {
          message: '',
          enabled: false,
        },
      },
      logging: {
        ...policy.linux.logging
      },
    },
  };
};

```

This way, we would still spread over specific fields, but new fields on the root would need to be manually added, and that would enforce the addition of new protection fields to the policy into this function.

Reply:
That was my biggest question, too: how to ensure that we won't forget to add new protections to `disableProtections`?

Actually, it is enforced: I added a `PolicyConfig` constant to the tests that will throw a type error if `PolicyConfig` is ever modified. See it [here](https://github.com/elastic/kibana/pull/144087/files#diff-5ce15bbb69cc01a413f019e5a8440212e99197da216b9a2218c794b949765820R73-R74).

Why did I add it to the test? Mostly because I didn't want to interfere with the current logic. Now there is only one default config (all ON). And there are 3 modifiers: one for stripping of paid features, one for enabling the support for paid features, and the new one for disabling protections.

I think we can go with either having a constant config for every scenario, or having these modifiers that can be applied after each other. As subsets of the config are independent from each other (like events from protections), I think using modifiers is a better choice.

This means, that `getCloudPolicyConfig()` can reuse `disableProtections()` right now, because it only disables protections and does not touch anything else.

What do you think? Would you like me to add `disableProtections()` to `getCloudPolicyConfig()`?

Reply:
Hey, @gergoabraham just tested adding new protection and the test works great, nicely done! 🙌

![image](https://user-images.githubusercontent.com/19270322/199317195-c71bac95-1bb8-4e8a-bd46-3fa677537be0.png)

Reply:
> What do you think? Would you like me to add `disableProtections()` to `getCloudPolicyConfig()`?

That makes sense to me, `getCloudPolicyConfig` inherits all policy defaults, then needs all protections off, and has `policy.linux.events.session_data` as `true`.  If you can make that change that would be awesome!